contracts/tec/contracts/src/MixinTECApprovalVerifier.sol

```diff
@@ -0,0 +1,200 @@
+        if (orders.length > 0) {
+            // Revert if approval is invalid for transaction orders
+            assertValidTransactionOrdersApproval(
+                transaction,
+                orders,
+                transactionSignature,
+                approvalExpirationTimeSeconds,
+                approvalSignatures
+            );
+        }
+    }
+
+    /// @dev Validates that the feeRecipients of a batch of order have approved a 0x transaction.
+    /// @param transaction 0x transaction containing salt, signerAddress, and data.
+    /// @param orders Array of order structs containing order specifications.
+    /// @param transactionSignature Proof that the transaction has been signed by the signer.
+    /// @param approvalExpirationTimeSeconds Array of expiration times in seconds for which each corresponding approval signature expires.
+    /// @param approvalSignatures Array of signatures that correspond to the feeRecipients of each order.
+    function assertValidTransactionOrdersApproval(
+        LibZeroExTransaction.ZeroExTransaction memory transaction,
+        LibOrder.Order[] memory orders,
+        bytes memory transactionSignature,
+        uint256[] memory approvalExpirationTimeSeconds,
+        bytes[] memory approvalSignatures
+    )
+        public
+        view
+    {
+        // Hash 0x transaction
+        bytes32 transactionHash = getTransactionHash(transaction);
+
+        // Create empty list of approval signers
+        address[] memory approvalSignerAddresses = new address[](0);
+
+        uint256 signaturesLength = approvalSignatures.length;
+        for (uint256 i = 0; i < signaturesLength; i++) {
+            // Create approval message
+            uint256 currentApprovalExpirationTimeseconds = approvalExpirationTimeSeconds[i];
+            TECApproval memory approval = TECApproval({
+                transactionHash: transactionHash,
+                transactionSignature: transactionSignature,
+                approvalExpirationTimeSeconds: currentApprovalExpirationTimeseconds
+            });
+
+            // Ensure approval has not expired
+            require(
+                // solhint-disable-next-line not-rely-on-time
+                currentApprovalExpirationTimeseconds > block.timestamp,
+                "APPROVAL_EXPIRED"
+            );
+
+            // Hash approval message and recover signer address
+            bytes32 approvalHash = getTECApprovalHash(approval);
+            address approvalSignerAddress = getSignerAddress(approvalHash, approvalSignatures[i]);
+    
+            // Add approval signer to list of signers
+            approvalSignerAddresses = approvalSignerAddresses.append(approvalSignerAddress);
+        }
+        
+        uint256 ordersLength = orders.length;
+        for (uint256 i = 0; i < ordersLength; i++) {
+            // Do not check approval if the order's senderAddress is null
+            if (orders[i].senderAddress == address(0)) {
```

Comment:
Shouldn't this check if the `senderAddress` _is_ the TEC contract? There are three cases:
1. SenderAddress is `null` -> fill however you want
2. SenderAddress is TECExtension -> Must fill through TEC contract
3. SenderAddress is otherExtension -> Which might or might not fill all orders via the TEC extension contract

We can't enforce 3) will always send orders through the TEC contract, so can we even support that use-case?

Reply:
Basically, this contract lets all of the standard 0x transaction failures fall back on the Exchange contract. Any `senderAddress` that isn't null or the TEC contract address is guaranteed to fail anyways, and it increases gas costs to check if it is the address of the TEC contract.